source/linker.cpp

```diff
@@ -0,0 +1,1098 @@
+      }
+    }
+    if (j == length)
+      prev_j = j;
+
+    std::memcpy(merged_binary + merged_length, binary + index,
+                (prev_j - index) * sizeof(spirv_word));
+    merged_length += prev_j - index;
+    indices[i] = j;
+  }
+
+  // Skip function declarations
+  for (uint32_t i = 0u; i < num_binaries; ++i) {
+    const uint32_t* const binary = binaries[i].data();
+    const size_t length = binaries[i].size();
+    const uint32_t index = indices[i];
+    uint32_t j = index;
+    uint32_t function_begin = j;
+    while (j < length) {
+      uint32_t word_count;
+      const spv::Op opcode = ParseOpcode(binary, j, word_count);
+
+      if (opcode == spv::Op::OpFunction)
+        function_begin = j;
+
+      // This was not a prototype: revert to the beginning of the function
+      if (opcode != spv::Op::OpFunction && opcode != spv::Op::OpFunctionParameter &&
+          opcode != spv::Op::OpFunctionEnd) {
+         j = function_begin;
+        break;
+      }
+
+      j += word_count;
+    }
+
+    indices[i] = j;
+  }
+
+  // Copy function defintions
+  for (uint32_t i = 0u; i < num_binaries; ++i) {
+    const uint32_t* const binary = binaries[i].data();
+    const size_t length = binaries[i].size();
+    const uint32_t index = indices[i];
+    std::memcpy(merged_binary + merged_length, binary + index,
+                (length - index) * sizeof(spirv_word));
+    merged_length += length - index;
+  }
+
+  // Link everything together
+  return TransformBinary(context, merged_spvbinary,
+                         [&link_table](uint32_t* binary, size_t word_offset){
+    spv::Id old_id = GetWord(binary, word_offset);
+    const auto link_iter = link_table.find(old_id);
+    if (link_iter != link_table.end())
+      SetWord(binary, word_offset, link_iter->second);
+  });
+}
+
+} // anonymous namespace
+
+spv_result_t spvLink(const spv_const_context context,
+                     const spv_const_binary binaries,
+                     const size_t num_binaries,
+                     spv_binary* linked_binary,
+                     spv_diagnostic* diagnostic) {
+  return spvLinkWithOptions(context, SPV_LINKER_OPTION_NONE, binaries,
+                            num_binaries, linked_binary, diagnostic);
+}
+
+spv_result_t spvLinkWithOptions(const spv_const_context context,
+                                const uint32_t options,
+                                const spv_const_binary binaries,
+                                const size_t num_binaries,
+                                spv_binary* linked_binary,
+                                spv_diagnostic* diagnostic) {
+  spv_context_t hijack_context = *context;
+  if (diagnostic) {
+    *diagnostic = nullptr;
+    libspirv::UseDiagnosticAsMessageConsumer(&hijack_context, diagnostic);
+  }
+
+  size_t total_word_count = 0u;
+  uint32_t id_upper_bound = 0u;
+  spv_result_t ret;
+  uint32_t max_version = 0u;
+
+  // As we need to bump IDs in each module, so as they do not conflict as we
```

Comment:
Please give a high level description.  
Phase 1: Shift the Ids in each binary so they occupy disjoint ranges.  In fact, I recommend you refactor out this transform step into its own function, so the high level structure is easier to see.
Phase 2: Merge the binaries.